core/src/read.jl

```diff
@@ -635,111 +635,143 @@ function PidControl(db::DB, config::Config, chunk_sizes::Vector{Int})::PidContro
     )
 end
 
-function UserDemand(db::DB, config::Config)::UserDemand
-    static = load_structvector(db, config, UserDemandStaticV1)
-    time = load_structvector(db, config, UserDemandTimeV1)
-
-    static_node_ids, time_node_ids, node_ids, valid =
-        static_and_time_node_ids(db, static, time, "UserDemand")
+function user_demand_static!(
+    active::BitVector,
+    demand::Vector{Float64},
+    demand_itp::Vector{Vector{ScalarInterpolation}},
+    return_factor::Vector{Float64},
+    min_level::Vector{Float64},
+    static::StructVector{UserDemandStaticV1},
+    node_ids::Vector{NodeID},
+    priorities::Vector{Int32},
+)::Nothing
+    user_demand_idx = 1
+    for group in IterTools.groupby(row -> row.node_id, static)
+        first_row = first(group)
+        while node_ids[user_demand_idx] < NodeID(NodeType.UserDemand, first_row.node_id)
+            user_demand_idx += 1
+        end
 
-    time_node_id_vec = NodeID.(NodeType.UserDemand, time.node_id)
+        active[user_demand_idx] = coalesce(first_row.active, true)
+        return_factor[user_demand_idx] = first_row.return_factor
+        min_level[user_demand_idx] = first_row.min_level
 
-    if !valid
-        error("Problems encountered when parsing UserDemand static and time node IDs.")
+        priority_idx = 1
+        for row in group
+            while priorities[priority_idx] < row.priority
+                priority_idx += 1
+            end
+            demand_itp[user_demand_idx][priority_idx].u .= row.demand
+            demand[(user_demand_idx - 1) * length(priorities) + priority_idx] = row.demand
+        end
     end
+    return nothing
+end
 
-    # All priorities used in the model
-    priorities = get_all_priorities(db, config)
-
-    active = BitVector()
-    min_level = Float64[]
-    return_factor = Float64[]
-    demand_itp = Vector{ScalarInterpolation}[]
-
+function user_demand_time!(
+    active::BitVector,
+    demand::Vector{Float64},
+    demand_itp::Vector{Vector{ScalarInterpolation}},
+    demand_from_timeseries::BitVector,
+    return_factor::Vector{Float64},
+    min_level::Vector{Float64},
+    time::StructVector{UserDemandTimeV1},
+    node_ids::Vector{NodeID},
+    priorities::Vector{Int32},
+    config::Config,
+)::Bool
     errors = false
-    trivial_timespan = [nextfloat(-Inf), prevfloat(Inf)]
     t_end = seconds_since(config.endtime, config.starttime)
+    user_demand_idx = 1
```

Comment:
Also here, it becomes a lot more readable if you pre-search the indices, then zip over them.

Something like this:

```julia
grouped = IterTools.groupby(row -> (row.node_id, row.priority), time)
group_ids = [first(group).node_id for group in grouped]
indices = searchsorted(node_ids, group_ids)

for (user_demand_idx, group) in zip(indices, grouped)
```